Cloning here for new code review workflow

thekuuzzz opened the pull request:
Just migrating these over, added try catches and made date a different function. Fixed strange spacing issues

thekuuzzz committed:
Cloning here for new code review workflow

Just migrating these over

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+
+
+def main():
+    if not os.path.exists('Preferences'):
```
remember not to have absolute paths, we wanted to have argparse provide you with a CLI to provide the path

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+def main():
+    if not os.path.exists('Preferences'):
+            print("Not a valid path")
+            sys.exit
```
good to have sys exit with a non zero value.. sys.exit(1) for instance

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            print("Not a valid path")
+            sys.exit
+    else:
+        with open('Preferences', 'rb') as fh:
```
again, remember to take in the name of the file from argparse

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            sys.exit
+    else:
+        with open('Preferences', 'rb') as fh:
+            json_object = json.load(fh)
```
I would add try/except here... in case the JSON fails to load

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            ChromePreferenceParser.parsePrint(json_object)
+
+
+class ChromePreferenceParser(object):
```
please add a docstring

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+class ChromePreferenceParser(object):
+
+    def _findIn(searchSpace, item):
+        # Recursive search for items inside a nested dictionary
```
have this as a docstring, not as a comment

kiddinn replied:
also search_space vs. searchSpace and function as _FindIn vs _findIn

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+
+    def _findIn(searchSpace, item):
+        # Recursive search for items inside a nested dictionary
+        foundItem = []
```
found_item vs foundItem, etc, this throughout (style guide)

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        # Recursive search for items inside a nested dictionary
+        foundItem = []
+        try:
+            for c, d in searchSpace.items():
```
also better instead of c, d to name the variables something a bit more descriptive...

what is c? and what is d? a key/value pair that contains?

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+                                    if isinstance(nestedItems, dict):
+                                        extraItems = findIn(nestedItems, item)
+                                        for evenMoreItems in extraItems:
+                                            foundItem.append(extraItems)
```
this is a really nested sentence here ... very hard to read really due to indentation...

one way of doing

```
if c == item:
  # something
else: 
  # something else
```

is to do

```
if c == item:
  found_item.append(d)
  continue

# something else
```

that saves you at least one indentation.... same for other places, for instance you've got

```
if isinstance....:
  nest_result = FindIn(d, item)
  for result....:
    # something
  else:

```
The else here applies to the for result... that's very confusing... this has to be fixed

and you've got two checks for isinstance dict...., sorry three checks for a dict... this sounds like something a recursive function could do easier and be more readable than this.... just check once if it is a dict and then call the recursive function in that case

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+                                        for evenMoreItems in extraItems:
+                                            foundItem.append(extraItems)
+            return foundItem
+        except Exception as e:
```
so a) what exception do you expect here? you are simply traversing a dict object to populate a list... what exception could there be?

b) don't catch Exception, catch something more narrow.... KeyError or what exceptions you may except, also you have a fairly large code block in this try/except, try to narrow it down to the actual code that could lead to an error

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        except Exception as e:
+            print("Exception running application, here's what we know")
+            print(e)
+            return 0
```
ok... so this is also confusing.... you return a list... except if an error comes up, then you return a 0, which is an integer.... that's very confusing for those that are calling this function, perhaps better to return an empty list? [] or?

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            return 0
+
+    def _dateConversion(webkitTimestamp):
+        # Converts date from webkit time to ISO 8601 time
```
again... _DateConversion

and webkit_timestamp

and start_date, etc, etc, etc

and also use a docstring, not a comment...

```
def _DateConversion(webkit_timestamp):
  """Converts date from ....

```

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        startDate = datetime.datetime(1601, 1, 1, 0, 0, 0, 0)
+        currentDate = datetime.timedelta(microseconds=int(webkitTimestamp))
+        isoTime = startDate + currentDate
+        return isoTime
```
in this case... why create iso_time?? why not just

`return start_date + current_date`

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+
+    def parseName():
+        # Parse through and get the account name, email, and profile name
+        # Returns the name, email and profile name
```
in docstring this would be...

```
def ParseName(json_object):
  """Parse through and return back account name....

  Args:
    json_object: some description here

  Returns:
    a tuple containing name, email and profile name as strings.
  """
  # Here starts the code...
```

also you don't ever take in the json_object into this function, yet you use it... meaning you are trying to use a global variable here, let's avoid that at all costs

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+                    profile = c.get('name')
+
+                    return name, email, profile
+        except Exception as e:
```
again... what exception are you expecting? and where?

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        # Returns the name, email and profile name
+        try:
+
+            for d in json_object['account_info']:
```
I would rather do....

```
account_info = json_object.get('account_info', None)

if account_info is None:
  return '', '', ''

for account in account_info:
  name = ....
```
that is the most likely cause for an error in this code is that json_object doesn't have hte key "account_info", you'll get a key error here, you won't get any error in this code if you do it like I wrote above and no need in this case to have error handling

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            raise EnvironmentError.filename
+            sys.exit
+
+    def parseDate():
```
same comments as before regarding docstring, naming convention and please please please please don't use global variables, pass json_object into the function

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        # Returns the last update date in ISO 8601 UTC
+        try:
+
+            webKitTime = json_object['last_update_date']
```
I would also use .get('last_update_date') 

and thus avoid the potential KeyError and need for error handling on this particular one, unless you expect other errors to come up, like you may get from _DateConversion (if the date is improperly formatted) and there you are casting a string to an int... so one error you may get from that code is simple to test....

```
$ ipython
In [1]: int('not a number')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-cd39f7b9e207> in <module>()
----> 1 int('not a number')

ValueError: invalid literal for int() with base 10: 'not a number'

```

You see.. you will get a ValueError from that....

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+    def parseMedia():
+        # Parses the following information from the media_engagement key
+        # 1: List of Websites 2: last modified 3. Media played? 4. visitcount
+        # returns these keys in a dict
```
see all other comments about naming + docstring

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        # returns these keys in a dict
+        try:
+            media = {}
+            for c, d in json_object['profile']['content_settings']['exceptions']['media_engagement'].iteritems():
```
here you are making a lot of assumptions....

json_object contains the key profile, which contains the key content_settings that contains the key exceptions....

so this needs a try/except or better yet have a better .get checks here to make sure this exists...

```
profile = json_object.get('profile', None)
if profile is None:
  # error out here

content_settings = profile.get('....

```
you get the picture

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        try:
+            media = {}
+            for c, d in json_object['profile']['content_settings']['exceptions']['media_engagement'].iteritems():
+                media.append(prefs.items(website=c, lastModified=dateConversion(d['last_modified']), playback=mediaPlaybacks, visitCount=visits))
```
way too long of a line, 

```
media.append(
    prefs......
    ....)
```

kiddinn replied:
what is prefs.items? I don't see that anywhere.... don't tell me that's another global variable?

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        # Gets a list of the current installed extensions and last update time
+        # returns a list of extension ID's and the last update time in ISO 8601
+
+        webKitTime = prefs.get(json_object['extensions']['autoupdate']['last_check'])
```
line too long

also doing three levels deep on the json_object.... let's check this better to make sure it's there

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+            for c in new_json:
+                printerList.append(c.get('selectedDestinationId'))
+                return printerList
+        except Exception as e:
```
same as usual with the error handling, see the style guide to how to handle errors

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+                printerList.append(c.get('selectedDestinationId'))
+                return printerList
+        except Exception as e:
+            raise parseException
```
and when you raise an exception you create you should have

```
class Error(Exception):
   """Base error class."""

class ParseException(Error):
   """Error raised when parser fails."""
```

and then you can do `raise ParseException('This is the reason why its raised')`

kiddinn commented on thekuuzzz/PrefParser.py:
```diff
@@ -0,0 +1,139 @@
+        print("JSON parsing failed")
+
+    if __name__ == "__main__":
+            main()
```
here I would like to see argparse, extracting command lines

kiddinn commented on thekuuzzz/argParse.py:
```diff
@@ -0,0 +1,22 @@
+from __future__ import absolute_import
```
I don't see this used anywhere

kiddinn reviewed (changes requested):
First round of reviews

thekuuzzz committed:
Added Sqlite Parser

Sqlite parser added
Json parser updated but not completed, will complete as soon as
possible.

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from urlparse import urlsplit
```
this should be after the stdlib imports

also just have this as

```
import urlsplit
```

see style guide

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+class QuarantineParse(object):
+    """Parses the QuarantineEventsV2 file and prints domains + download agents
+
+        Args:
```
skip these args and returns section for a class... not needed, shouldn't be here.. also not used

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+
+
+class QuarantineParse(object):
+    """Parses the QuarantineEventsV2 file and prints domains + download agents
```
this should probably not be done by the same object....

I would say that this class should be the parser, as the name implies.. and the parser returns back strings...

then you've got somewhere a __main__ section that accepts the input from the user, parses the file using this class and sends it to a print function that takes care of the printing.

So this should rather be something like
```
"""Parser for the QuarantineEventsV2 SQLite database."""
```

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+        none
+    """
+
+    def parseFile():
```
where are the arguments here?

this implies (which gets later confirmed) that the path is hardcoded.. should not be the case here... you need to have this accept an argument, the argument being the filename/path to the file to be parsed.

kiddinn replied:
and you are also missing a docstring ... and **self** as the first parameter to the function since this is a class function

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+
+    def parseFile():
+        agent_dict = dict(agent=' ', count=0)
+        url_dict = dict(site=' ', count=0)
```
take a look at `collections.Counter` ... I would highly encourage you to use that instead of trying to implement your own counter

kiddinn replied:
and this should be `url_counter` instead of `url_dict` ;)

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+        agent_dict = dict(agent=' ', count=0)
+        url_dict = dict(site=' ', count=0)
+        try:
+            database = sqlite3.connect('QuarantineEventsV2')
```
why is the entire codebase here inside the try/except loop?

you should try to limit the try/except clause to just the code that really could raise the exception...

there is no reason to believe that splitting a database query record into a list would raise a EnvironmentError...

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+        url_dict = dict(site=' ', count=0)
+        try:
+            database = sqlite3.connect('QuarantineEventsV2')
+            database_query = database.execute('SELECT LSQuarantineTimestamp AS Time,
```
split this into two parts, also line breaks and length....

```
database_query = (
    'SELECT LSQ.....'
    'LSQuar..... AS DATA')

database_results = database.execute(database_query)
```

don't do it all in one piece... makes it easier to read over, debug, etc.

kiddinn replied:
also in this case since you are using a class... this query of yours does not change... so I would recommend it to be a constant saved as s class constant..

```
class QuarantineParser(object):
   """Docstring and...."""

  QUERY = (
      'SELECT....
```

That way you can just reference this here as

```database_results = database.execute(self.QUERY)```

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            LSQuarantineAgentName AS Agent, LSQuarantineOriginURLString AS URL,
+            LSQuarantineDataURLString AS Data
+            FROM LSQuarantineEvent ORDER BY Time;')
+            for record in database_query.split('|'):
```
this is also very confusing....

why are there many records in the `database_query`????

see this is why it is confusing when you name variables something they are not... this is not a database query, this is the actual result set that you get from executing a query in the database... name it accordingly...

also please do something like

```
for record in result:
   # this gets you all the records not just a single one
```

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            LSQuarantineDataURLString AS Data
+            FROM LSQuarantineEvent ORDER BY Time;')
+            for record in database_query.split('|'):
+                if record.startswith(str.isdigit()):
```
this is also a very weird parsing here... you get a row back...

```
for row in result:
   time = row[0]
```
you can even make the row not be a list but be a dict by changing the row factory

```database.row_factory = sqlite3.Row```

This way the row would have something like `row.get('Time')` in your query above... or just `row['Time']` (since you used *LSQuarantineTimestamp AS Time*)

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            for record in database_query.split('|'):
+                if record.startswith(str.isdigit()):
+                    time = _convert_time(record)
+                elif record.startswith('https') | record.startswith('http'):
```
again, this is just not good parsing here... use the row

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+                        else:
+                            agent_dict[record] = agent_dict[record]+1
+                            source = record
+            print(time + ' ' + source + ' downloaded from ' + url + '\n')
```
let's keep printing out of this class and don't concatenate like this rather use

`'{0:S} {1:s} downloaded from {2:s}.format(time, source, url)`

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            print('\n' + 'List of Domains: ' + '\n')
+            for domains, values in url_dict:
+                print(domains + ' -> ' + values)
+        except EnvironmentError as e:
```
where are all these errors coming from? where are they raised... keep them close by...

what about sqlite3.DatabaseError ? I don't see that caught here...

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            print('I/O error with database file' + e)
+        except EOFError as eof:
+            print('Parser reached end of file and didnt expect it' + eof)
+        except Exception as eek:
```
don't do this....

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+                print(domains + ' -> ' + values)
+        except EnvironmentError as e:
+            print('I/O error with database file' + e)
+        except EOFError as eof:
```
why are you catching EOFError?? what would possibly raise that

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+    Returns:
+     ISO 8601 timestamp
+
+    """
```
keep this as a docstring... not outside of the function

```
def _convert_time(**self**, cocoa_time):
  """Helper method....
```

also remember the first argument should be `self` here since this is a class method

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+    """
+    def _convert_time(cocoa_time):
+        try:
+            start_time = datetime.datetime(2001, 1, 1, 0, 0, 0, 0)
```
you are constantly calculating this, again and again... for each timestamp.... this does not change... and therefore it should be a CONSTANT, preferably here in this case defined as a class constant.... and preferably just calculated manually once by you and stored as an int

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+    """ Helper method to convert from cocoa to ISO time
+
+    Args:
+     cocoa timestamp
```
`cocoa_time (str): the timestamp in cocoa format`

or is it an int when you read it properly using the row

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+    def _convert_time(cocoa_time):
+        try:
+            start_time = datetime.datetime(2001, 1, 1, 0, 0, 0, 0)
+            current_time = datetime.timedelta(microseconds=int(cocoa_time))
```
don't do these many operations at once... for instnace you are taking a string `cocoa_time` and converting it to an int, which can raise an exception at the same time as you are doing a timedelta.... let's rather do
```
try:
  cocoa_time_int = int(cocoa_time, 10)
except ValueError:
  # do something here
current_time = datetime.timedelta(microseconds=cocoa_time_int)
```

that is keep these separated

kiddinn commented on thekuuzzz/sqlite/QuarantineParse.py:
```diff
@@ -0,0 +1,78 @@
+            new_time = start_time + current_time
+            new_time = datetime.datetime.isoformat(new_time)
+            return new_time
+        except exception as e:
```
this should nt happen... don't expect a generic exception here

kiddinn reviewed (changes requested):
first round done for the sqlite parser
